ADBDEV-3674-new: new shouldn't return NULL

RekGRpth opened the pull request:
ORCA makes extensive use of the [GPOS_NEW](https://github.com/arenadata/gpdb/blob/88532d712d0258ab14e5992854bd04605d1e763f/src/backend/gporca/libgpos/include/gpos/memory/CMemoryPool.h#LL299C9-L299C17) to allocate memory, not checking the result for NULL, but instead relying on throwing an exception.
[The implementation of this macro](https://github.com/arenadata/gpdb/blob/88532d712d0258ab14e5992854bd04605d1e763f/src/backend/gporca/libgpos/include/gpos/memory/CMemoryPool.h#L276-L281) depends on the memory pool.
Depending on the variable [optimizer_use_gpdb_allocators](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gpopt/CGPOptimizer.cpp#L174-L177) can be used [either](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gpopt/utils/CMemoryPoolPallocManager.cpp#L53-L58) GreenPlum's memory pool manager [CMemoryPoolPalloc](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gpopt/utils/CMemoryPoolPalloc.cpp#L33-L61) (default), [or](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/memory/CMemoryPoolManager.cpp#L69-L79) ORCA's memory pool manager [CMemoryPoolTracker](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/memory/CMemoryPoolTracker.cpp#L67-L118).
When there is not enough memory for the GreenPlum's memory pool manager in the [gpdb::GPDBMemoryContextAlloc](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gpopt/gpdbwrappers.cpp#L2689-L2698) function an exception is thrown using the macro [GP_WRAP_END](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gpopt/gpdbwrappers.cpp#L44-L50) and therefore the lines, [testing for NULL](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gpopt/utils/CMemoryPoolPalloc.cpp#L50-L53) are never run and can be safely removed.
And if there is not enough memory for the ORCA's memory pool manager in the [CMemoryPoolTracker::NewImpl](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/memory/CMemoryPoolTracker.cpp#L67-L118) function an exception is thrown using the macro [GPOS_OOM_CHECK](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/memory/CMemoryPoolTracker.cpp#LL90C22-L90C22), but for some reason [before that, the result is checked for NULL and NULL is returned](https://github.com/arenadata/gpdb/blob/dfe812f45701096a07a76c341ad8a92b3315555b/src/backend/gporca/libgpos/src/memory/CMemoryPoolTracker.cpp#L85-L88), which can lead to a segfault with further use of the NULL result, so the rows must be deleted.
Unfortunately there is no easy way to test this. We came up with a couple of options, but both of them are not very good: either make a fault-injector, or replace the allocator through preload.

RekGRpth commented:
I'll update the commit description if that description is clear.

HustonMmmavr commented on src/backend/gpopt/utils/CMemoryPoolPalloc.cpp:
```diff
@@ -47,11 +47,6 @@ CMemoryPoolPalloc::NewImpl(const ULONG bytes, const CHAR *, const ULONG,
-
 		SArrayAllocHeader *header = static_cast<SArrayAllocHeader *>(ptr);
 
 		header->m_user_size = bytes;
```
Seems here would be a `NULL`-pointer dereference, if ptr is `NULL`. Maybe it's worth to add `GPOS_OOM_CHECK(ptr)` like at `CMemoryPoolTracker`?

HustonMmmavr replied:
Or `GPDBMemoryContextAlloc` will raise an error? If so, maybe it's worth to add comment about that

RekGRpth replied:
> Seems here would be a `NULL`-pointer dereference, if ptr is `NULL`.

No! In this case, GPDBMemoryContextAlloc throws an exception! And I mentioned it in the description.

HustonMmmavr replied:
Sorry, I've forgot to submit review, so this message

> Or GPDBMemoryContextAlloc will raise an error? If so, maybe it's worth to add comment about that

wasn't visible

InnerLife0 commented:
> > There is no easy way of testing this
> 
> Yes, we came up with a couple of options, but both of them are not very good: either make a fault-injector, or replace the allocator through preload.

Thanks.
Overall looks OK, but we should add a few words about this in description.

Nitpicking on writing, there is some logical inconsistency in existed description. Functions shouldn't return NULL not because of existed exception, but because of their contract, their API. Exception is the way of errors handling and used to confirm our contract.

RekGRpth commented:
> there is some logical inconsistency in existed description

I've updated the pull request description.

InnerLife0 commented:
> I've updated the pull request description.

Thanks.
What about my first suggestion about adding few words to describe a complexity of testing? If you find some suggestions redundant it's OK to say.

RekGRpth commented:
I've updated the pull request description again.

RekGRpth committed:
new shouldn't return NULL
ORCA makes extensive use of the GPOS_NEW to allocate memory,
not checking the result for NULL, but instead relying on throwing an exception.
The implementation of this macro depends on the memory pool.
Depending on the variable optimizer_use_gpdb_allocators can be used either
GreenPlum's memory pool manager CMemoryPoolPalloc (default), or
ORCA's memory pool manager CMemoryPoolTracker.
When there is not enough memory for the GreenPlum's memory pool manager
in the gpdb::GPDBMemoryContextAlloc function an exception is thrown using
the macro GP_WRAP_END and therefore the lines, testing for NULL are never
run and can be safely removed.
And if there is not enough memory for the ORCA's memory pool manager
in the CMemoryPoolTracker::NewImpl function an exception is thrown using
the macro GPOS_OOM_CHECK, but for some reason before that,
the result is checked for NULL and NULL is returned, which can lead to
a segfault with further use of the NULL result, so the rows must be deleted.
Unfortunately there is no easy way to test this.
We came up with a couple of options, but both of them are not very good:
either make a fault-injector, or replace the allocator through preload.

RekGRpth committed:
Merge branch 'adb-6.x-dev' into ADBDEV-3674-new

Stolb27 committed:
Merge branch 'adb-6.x-dev' into ADBDEV-3674-new
